src/ifcpatch/ifcpatch/recipes/AssignConstituentFractions.py

```diff
@@ -0,0 +1,151 @@
+                constituent_set_name = constituent_set.Name or "Unnamed Constituent Set"
+                self.logger.warning(f"No widths found for constituents in set '{constituent_set_name}'. Skipping.")
+                continue
+
+            # Assign fractions based on widths
+            for constituent, width in constituent_widths.items():
+                fraction = width / total_width
+                constituent.Fraction = fraction
+                self.logger.info(
+                    f"Constituent: {constituent.Name}, "
+                    f"Width: {width:.4f} {length_unit}, "
+                    f"Fraction: {fraction:.4f}"
+                )
+
+    def get_element_quantities(self, element: ifcopenshell.entity_instance) -> Dict[str, float]:
+        """Get width quantities for an element."""
+        quantities = {}
+        
+        # Get quantities from IfcPhysicalComplexQuantity
+        for rel in getattr(element, 'IsDefinedBy', []):
+            if not rel.is_a('IfcRelDefinesByProperties'):
+                continue
+            definition = rel.RelatingPropertyDefinition
+            if not definition.is_a('IfcElementQuantity'):
+                continue
+            for quantity in definition.Quantities:
+                if not quantity.is_a('IfcPhysicalComplexQuantity'):
+                    continue
+                if quantity.Discrimination.lower() != 'layer':
+                    continue
+                for sub_quantity in quantity.HasQuantities:
+                    if not sub_quantity.is_a('IfcQuantityLength'):
+                        continue
+                    if sub_quantity.Name.lower() != 'width':
+                        continue
+                    quantities[quantity.Name.lower()] = float(sub_quantity.LengthValue)
+                
+        return quantities
+
+    def calculate_constituent_widths(
+        self,
+        constituents: List[ifcopenshell.entity_instance],
+        elements: set[ifcopenshell.entity_instance],
+        unit_scale: float
+    ) -> Tuple[Dict[ifcopenshell.entity_instance, float], float]:
+        """Calculate the widths of constituents based on associated quantities."""
+        constituents_by_name = defaultdict(list)
+        for constituent in constituents:
+            constituent_name = (constituent.Name or "Unnamed Constituent").strip().lower()
+            constituents_by_name[constituent_name].append(constituent)
+
+        # Collect all quantities from all elements
+        element_quantities = {}
+        for element in elements:
+            element_quantities.update(self.get_element_quantities(element))
+
+        constituent_widths = {}
+        total_width = 0.0
+
+        for constituent_name, constituents_list in constituents_by_name.items():
+            width = element_quantities.get(constituent_name, 0.0)
+            for constituent in constituents_list:
+                constituent_widths[constituent] = width
+                total_width += width
```

Comment:
I think you should reverse this loop, see this explanation:

![2025-01-16-140856_1168x768_scrot](https://github.com/user-attachments/assets/9a5f6a1e-535d-4319-8ac2-0d1b27bda5df)